examples/layer-surface.c

```diff
@@ -0,0 +1,434 @@
+	if (eglInitialize(egl_display, &major, &minor) != EGL_TRUE) {
+		fprintf(stderr, "Can't initialise egl display\n");
+		exit(EXIT_FAILURE);
+	}
+	printf("EGL major: %d, minor %d\n", major, minor);
+
+	if (!eglBindAPI(EGL_OPENGL_ES_API)) {
+		fprintf(stderr, "Can't bind API\n");
+		exit(EXIT_FAILURE);
+	} else {
+		fprintf(stderr, "Bound API\n");
+	}
+
+	EGLint count;
+	eglGetConfigs(egl_display, NULL, 0, &count);
+	printf("EGL has %d configs\n", count);
+
+	EGLConfig *configs = calloc(count, sizeof(EGLConfig));
+	if (configs == NULL) {
+		fprintf(stderr, "Can't allocate EGL configs\n");
+		exit(EXIT_FAILURE);
+	}
+
+	EGLint n;
+	eglChooseConfig(egl_display, config_attribs, configs, count, &n);
+
+	EGLint size;
+	for (int i = 0; i < n; ++i) {
+		eglGetConfigAttrib(egl_display, configs[i], EGL_BUFFER_SIZE, &size);
+		printf("Buffer size for config %d is %d\n", i, size);
+		eglGetConfigAttrib(egl_display, configs[i], EGL_RED_SIZE, &size);
+		printf("Red size for config %d is %d\n", i, size);
+
+		egl_conf = configs[i];
+		break;
+	}
+
+	egl_context = eglCreateContext(egl_display, egl_conf, EGL_NO_CONTEXT,
+		context_attribs);
+}
+
+static void create_window() {
+	egl_window = wl_egl_window_create(surface, width, height);
+	if (egl_window == EGL_NO_SURFACE) {
+		fprintf(stderr, "Can't create egl window\n");
+		exit(EXIT_FAILURE);
+	} else {
+		fprintf(stderr, "Created egl window\n");
+	}
+
+	egl_surface = eglCreateWindowSurface(egl_display, egl_conf,
+		(EGLNativeWindowType)egl_window, NULL);
+
+	draw();
+
+	wl_display_dispatch(display);
+	wl_display_roundtrip(display);
+}
+
+enum surface_layers_layer parse_layer(const char *s) {
+	if (strcmp(s, "background") == 0) {
+		return SURFACE_LAYERS_LAYER_BACKGROUND;
+	} else if (strcmp(s, "bottom") == 0) {
+		return SURFACE_LAYERS_LAYER_BOTTOM;
+	} else if (strcmp(s, "top") == 0) {
+		return SURFACE_LAYERS_LAYER_TOP;
+	} else if (strcmp(s, "overlay") == 0) {
+		return SURFACE_LAYERS_LAYER_OVERLAY;
+	}
+	fprintf(stderr, "Unknown layer name: %s\n", s);
+	exit(EXIT_FAILURE);
+}
+
+enum layer_surface_anchor parse_anchor(const char *s) {
+	if (strcmp(s, "top") == 0) {
+		return LAYER_SURFACE_ANCHOR_TOP;
+	} else if (strcmp(s, "bottom") == 0) {
+		return LAYER_SURFACE_ANCHOR_BOTTOM;
+	} else if (strcmp(s, "left") == 0) {
+		return LAYER_SURFACE_ANCHOR_LEFT;
+	} else if (strcmp(s, "right") == 0) {
+		return LAYER_SURFACE_ANCHOR_RIGHT;
+	}
+	fprintf(stderr, "Unknown anchor name: %s\n", s);
+	exit(EXIT_FAILURE);
+}
+
+enum layer_surface_input_device parse_input_device(const char *s) {
+	if (strcmp(s, "pointer") == 0) {
+		return LAYER_SURFACE_INPUT_DEVICE_POINTER;
+	} else if (strcmp(s, "keyboard") == 0) {
+		return LAYER_SURFACE_INPUT_DEVICE_KEYBOARD;
+	} else if (strcmp(s, "touch") == 0) {
+		return LAYER_SURFACE_INPUT_DEVICE_TOUCH;
+	}
+	fprintf(stderr, "Unknown input device: %s\n", s);
+	exit(EXIT_FAILURE);
+}
+
+void parse_pair(const char *s, uint32_t *x, uint32_t *y) {
+	char *buf = strdup(s);
```

Comment:
Why are you duplicating (and then leaking) the string here?
You don't need to worry about the comma messing with atoi. It'll immediately stop once it encounters a non-digit.

Reply:
Actually, thinking about it, you can just use `sscanf(s, "%u,%u", x, y)`.

Reply:
-1 to sscanf. +1 to fixing the leak

Reply:
sscanf does literally the exact same thing with no unnecessary and actually allows errors to checked a bit better than the current implementation.
```c
errno = 0;
if (sscanf(s, "%u,%u", x, y) != 2 || errno != 0) {
    fprintf(stderr, "Invalid format\n");
    exit(EXIT_FAILURE);
}
```
I agree that scanf (the stdin one) is shit, but sscanf is fine.

Reply:
It works, but I don't really like it.